Unstable test: TestHotReloadUIConfigTempFile

I've seen this one failing sporadically:
```
=== RUN   TestHotReloadUIConfigTempFile
    static_handler_test.go:279: 
        	Error Trace:	static_handler_test.go:279
        	Error:      	"" does not contain "About a new Jaeger"
        	Test:       	TestHotReloadUIConfigTempFile
        	Messages:   	[{Entry:{Level:info Time:2022-05-25 20:44:07.707232055 +0000 UTC m=+2.677106569 LoggerName: Message:watching Caller:undefined Stack:} Context:[{Key:file Type:15 Integer:0 String:/tmp/ui-config-hotreload-225331457/486536771.json Interface:<nil>}]} {Entry:{Level:info Time:2022-05-25 20:44:07.707258755 +0000 UTC m=+2.677133369 LoggerName: Message:watching Caller:undefined Stack:} Context:[{Key:dir Type:15 Integer:0 String:/tmp/ui-config-hotreload-225331457 Interface:<nil>}]} {Entry:{Level:info Time:2022-05-25 20:44:07.708553363 +0000 UTC m=+2.678427877 LoggerName: Message:reloading UI config Caller:undefined Stack:} Context:[{Key:filename Type:15 Integer:0 String:/tmp/ui-config-hotreload-225331457/486536771.json Interface:<nil>}]} {Entry:{Level:error Time:2022-05-25 20:44:07.708720364 +0000 UTC m=+2.678594978 LoggerName: Message:error while reloading the UI config Caller:undefined Stack:} Context:[{Key:error Type:26 Integer:0 String: Interface:cannot parse UI config file /tmp/ui-config-hotreload-225331457/486536771.json: unexpected end of JSON input}]} {Entry:{Level:info Time:2022-05-25 20:44:07.708730865 +0000 UTC m=+2.678605379 LoggerName: Message:reloaded UI config Caller:undefined Stack:} Context:[{Key:filename Type:15 Integer:0 String:/tmp/ui-config-hotreload-225331457/486536771.json Interface:<nil>}]} {Entry:{Level:info Time:2022-05-25 20:44:07.708748465 +0000 UTC m=+2.678623079 LoggerName: Message:reloading UI config Caller:undefined Stack:} Context:[{Key:filename Type:15 Integer:0 String:/tmp/ui-config-hotreload-225331457/486536771.json Interface:<nil>}]} {Entry:{Level:error Time:2022-05-25 20:44:07.708856265 +0000 UTC m=+2.678730879 LoggerName: Message:error while reloading the UI config Caller:undefined Stack:} Context:[{Key:error Type:26 Integer:0 String: Interface:cannot parse UI config file /tmp/ui-config-hotreload-225331457/486536771.json: unexpected end of JSON input}]} {Entry:{Level:info Time:2022-05-25 20:44:07.708863765 +0000 UTC m=+2.678738379 LoggerName: Message:reloaded UI config Caller:undefined Stack:} Context:[{Key:filename Type:15 Integer:0 String:/tmp/ui-config-hotreload-225331457/486536771.json Interface:<nil>}]}]
--- FAIL: TestHotReloadUIConfigTempFile (0.02s)
```
Likely some kind of race condition dependent on the timing of hot reload.

--- a/cmd/query/app/static_handler_test.go
+++ b/cmd/query/app/static_handler_test.go
@@ -240,25 +240,30 @@ func TestWatcherError(t *testing.T) {
 	}
 }
 
-func TestHotReloadUIConfigTempFile(t *testing.T) {
+func TestHotReloadUIConfig(t *testing.T) {
 	dir, err := os.MkdirTemp("", "ui-config-hotreload-*")
 	require.NoError(t, err)
 	defer os.RemoveAll(dir)
 
-	tmpfile, err := os.CreateTemp(dir, "*.json")
+	cfgFile, err := os.CreateTemp(dir, "*.json")
 	require.NoError(t, err)
-	tmpFileName := tmpfile.Name()
+	defer cfgFile.Close()
+	cfgFileName := cfgFile.Name()
+
+	tmpFile, err := os.CreateTemp(dir, "*.json")
+	require.NoError(t, err)
+	defer tmpFile.Close()
 
 	content, err := os.ReadFile("fixture/ui-config-hotreload.json")
 	require.NoError(t, err)
 
-	err = syncWrite(tmpFileName, content, 0644)
+	err = syncWrite(cfgFile, tmpFile, content)
 	require.NoError(t, err)
 
 	zcore, logObserver := observer.New(zapcore.InfoLevel)
 	logger := zap.New(zcore)
 	h, err := NewStaticAssetsHandler("fixture", StaticAssetsHandlerOptions{
-		UIConfigPath: tmpFileName,
+		UIConfigPath: cfgFileName,
 		Logger:       logger,
 	})
 	require.NoError(t, err)
@@ -267,12 +272,12 @@ func TestHotReloadUIConfigTempFile(t *testing.T) {
 	assert.Contains(t, c, "About Jaeger")
 
 	newContent := strings.Replace(string(content), "About Jaeger", "About a new Jaeger", 1)
-	err = syncWrite(tmpFileName, []byte(newContent), 0644)
+	err = syncWrite(cfgFile, tmpFile, []byte(newContent))
 	require.NoError(t, err)
 
 	waitUntil(t, func() bool {
 		return logObserver.FilterMessage("reloaded UI config").
-			FilterField(zap.String("filename", tmpFileName)).Len() > 0
+			FilterField(zap.String("filename", cfgFileName)).Len() > 0
 	}, 100, 10*time.Millisecond, "timed out waiting for the hot reload to kick in")
 
 	i := string(h.indexHTML.Load().([]byte))
@@ -392,8 +397,8 @@ func waitUntil(t *testing.T, f func() bool, iterations int, sleepInterval time.D
 
 // syncWrite ensures data is written to the given filename and flushed to disk.
 // This ensures that any watchers looking for file system changes can be reliably alerted.
-func syncWrite(filename string, data []byte, perm os.FileMode) error {
-	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC|os.O_SYNC, perm)
+func syncWrite(target *os.File, temp *os.File, data []byte) error {
+	f, err := os.OpenFile(temp.Name(), os.O_WRONLY|os.O_CREATE|os.O_TRUNC|os.O_SYNC, 0644)
 	if err != nil {
 		return err
 	}
@@ -401,5 +406,8 @@ func syncWrite(filename string, data []byte, perm os.FileMode) error {
 	if _, err = f.Write(data); err != nil {
 		return err
 	}
-	return f.Sync()
+	if err := f.Sync(); err != nil {
+		return err
+	}
+	return os.Rename(temp.Name(), target.Name())
 }

